uniffi/src/ffi/handle_map.rs

```diff
@@ -0,0 +1,414 @@
+    ///
+    /// # Panics
+    ///
+    /// This will panic if a previous `get()` call has panicked
+    /// inside its callback.
+    ///
+    /// (It may also panic if the handle map detects internal state corruption,
+    /// however this should not happen except for bugs in the handle map code).
+    pub fn get<F, E, R>(&self, h: Handle, callback: F) -> Result<R, E>
+    where
+        F: FnOnce(&T) -> Result<R, E>,
+        E: From<HandleError>,
+    {
+        let obj = {
+            let map = self.map.read().unwrap();
+            let obj = map.get(h)?;
+            Arc::clone(&obj)
+        };
+        callback(&*obj)
+    }
+
+    /// Convenient wrapper for `get` which takes a `u64` that it will convert to
+    /// a handle.
+    ///
+    /// The other benefit (besides convenience) of this over the version
+    /// that takes a [`Handle`] is that it allows handling handle-related errors
+    /// in one place.
+    ///
+    /// This takes the map's `read` lock.
+    pub fn get_u64<F, E, R>(&self, u: u64, callback: F) -> Result<R, E>
+    where
+        F: FnOnce(&T) -> Result<R, E>,
+        E: From<HandleError>,
+    {
+        self.get(Handle::from_u64(u)?, callback)
+    }
+
+    /// Helper that performs both a [`call_with_result`] and [`get`](ArcHandleMap::get).
+    ///
+    /// This takes the map's `read` lock for as long as needed to clone the inner `Arc`.
+    /// This is so the lock isn't held while the callback is in use.
+    pub fn call_with_result<R, E, F>(
+        &self,
+        out_error: &mut ExternError,
+        h: u64,
+        callback: F,
+    ) -> R::Value
+    where
+        F: std::panic::UnwindSafe + FnOnce(&T) -> Result<R, E>,
+        ExternError: From<E>,
+        R: IntoFfi,
+    {
+        use ffi_support::call_with_result;
+        call_with_result(out_error, || -> Result<_, ExternError> {
+            // We can't reuse `get` here because it would require E:
+            // From<HandleError>, which is inconvenient...
+            let h = Handle::from_u64(h)?;
+            let obj = {
+                let map = self.map.read().unwrap();
+                let obj = map.get(h)?;
+                Arc::clone(&obj)
+            };
+            Ok(callback(&*obj)?)
+        })
+    }
+
+    /// Helper that performs both a [`call_with_output`] and [`get`](ArcHandleMap::get).
+    pub fn call_with_output<R, F>(
+        &self,
+        out_error: &mut ExternError,
+        h: u64,
+        callback: F,
+    ) -> R::Value
+    where
+        F: std::panic::UnwindSafe + FnOnce(&T) -> R,
+        R: IntoFfi,
+    {
+        self.call_with_result(out_error, h, |r| -> Result<_, HandleError> {
+            Ok(callback(r))
+        })
+    }
+
+    /// Use `constructor` to create and insert a `T`, while inside a
+    /// [`call_with_result`] call (to handle panics and map errors onto an
+    /// `ExternError`).
+    ///
+    /// This takes the map's `write` lock.
```

Comment:
nit: same comment here re locks - the write lock is not held while the ctor runs, which is probably worth explicitly noting.

(As I mentioned elsewhere, it might make sense to document the lock semantics on the struct itself and avoid duplicating it for every method call - best I can tell, the semantics are identical in all cases (other than read vs write locks)